src/gui/layertree/qgslayertreeview.cpp

```diff
@@ -43,6 +44,10 @@ QgsLayerTreeView::QgsLayerTreeView( QWidget *parent )
   setEditTriggers( EditKeyPressed );
   setExpandsOnDoubleClick( false ); // normally used for other actions
 
+  // Ensure legend graphics are scrollable
+  header()->setStretchLastSection( false );
+  header()->setSectionResizeMode( QHeaderView::ResizeToContents );
```

Comment:
Currently, the tree view column count is 1. The other [setSectionResizeMode call](https://doc.qt.io/qt-5/qheaderview.html#setSectionResizeMode-1), specifying column, can be used, if needed.